File: source/c/minimum_spanning_tree.c
// Importing the stdio.h and limits.h libraries for input/output and constants
#include <stdio.h>
#include <limits.h>

// Defining a constant for the number of vertices in the graph
#define V 5

// Defining a function to find the vertex with the minimum key value
int minKey(int key[], int mstSet[])
{
  // Initializing the minimum value and index
  int min = INT_MAX, min_index;
  // Looping through all the vertices
  for (int v = 0; v < V; v++)
    // If the vertex is not in the MST set and has a smaller key value than the current minimum
    if (mstSet[v] == 0 && key[v] < min)
      // Update the minimum value and index
      min = key[v], min_index = v;
  // Return the minimum index
  return min_index;
}

// Defining a function to print the MST and the edge weights
void printMST(int parent[], int graph[V][V])
{
  // Printing the header
  printf("Edge \tWeight\n");
  // Looping through all the vertices except the first one
  for (int i = 1; i < V; i++)
    // Printing the edge and the weight
    printf("%d - %d \t%d \n", parent[i], i, graph[i][parent[i]]);
}

// Defining a function to find the minimum spanning tree using Prim's algorithm
void primMST(int graph[V][V])
{
  // Initializing the arrays to store the key values, the MST set, and the parent vertices
  int key[V];
  int mstSet[V];
  int parent[V];
  // Assigning infinite values to all the key values and false values to all the MST set
  for (int i = 0; i < V; i++)
    key[i] = INT_MAX, mstSet[i] = 0;
  // Setting the key value of the first vertex to zero and the parent to -1
  key[0] = 0;
  parent[0] = -1;
  // Looping until all the vertices are included in the MST set
  for (int count = 0; count < V - 1; count++)
  {
    // Finding the vertex with the minimum key value
    int u = minKey(key, mstSet);
    // Adding the vertex to the MST set
    mstSet[u] = 1;
    // Updating the key values and parent vertices of the adjacent vertices
    for (int v = 0; v < V; v++)
      // If there is an edge between u and v, v is not in the MST set, and the weight of the edge is smaller than the current key value of v
      if (graph[u][v] && mstSet[v] == 0 && graph[u][v] < key[v])
        // Update the key value and the parent vertex of v
        parent[v] = u, key[v] = graph[u][v];
  }
  // Printing the MST and the edge weights
  printMST(parent, graph);
}

// Defining the main function
int main(int argc, char** argv)
{
  // Defining a sample graph as an adjacency matrix
  // The graph is the same as the one shown in [this article](https://www.programiz.com/dsa/prim-algorithm)
  int graph[V][V] = {{0, 2, 0, 6, 0},
                     {2, 0, 3, 8, 5},
                     {0, 3, 0, 0, 7},
                     {6, 8, 0, 0, 9},
                     {0, 5, 7, 9, 0}};
  // Calling the primMST function with the graph
  primMST(graph);
  // Returning zero to indicate successful termination
  return 0;
}

File: source/python/minimum_spanning_tree.py
# Importing the heapq module for priority queue
import heapq

# Defining a function to find the minimum spanning tree
def prim(graph, start):
  # Initializing the variables
  mst = [] # List to store the edges of the minimum spanning tree
  visited = set() # Set to keep track of the visited nodes
  pq = [] # Priority queue to store the edges with their weights
  # Adding the start node and its adjacent edges to the priority queue
  visited.add(start)
  for dest, weight in graph[start].items():
    heapq.heappush(pq, (weight, start, dest))
  # Looping until the priority queue is empty
  while pq:
    # Popping the edge with the minimum weight
    weight, src, dest = heapq.heappop(pq)
    # Checking if the destination node is already visited
    if dest in visited:
      continue
    # Adding the edge to the minimum spanning tree
    mst.append((src, dest, weight))
    # Marking the destination node as visited
    visited.add(dest)
    # Adding the destination node's adjacent edges to the priority queue
    for next_dest, next_weight in graph[dest].items():
      heapq.heappush(pq, (next_weight, dest, next_dest))
  # Returning the minimum spanning tree
  return mst

# Defining a sample graph as an adjacency list
# The graph is the same as the one shown in [this article](https://www.geeksforgeeks.org/prims-minimum-spanning-tree-mst-greedy-algo-5/)
graph = {
  "A": {"B": 2, "D": 3},
  "B": {"A": 2, "C": 4, "E": 3},
  "C": {"B": 4, "E": 1, "F": 5},
  "D": {"A": 3, "E": 4, "G": 6},
  "E": {"B": 3, "C": 1, "D": 4, "F": 2, "G": 5},
  "F": {"C": 5, "E": 2, "H": 3},
  "G": {"D": 6, "E": 5, "H": 4},
  "H": {"F": 3, "G": 4}
}

# Calling the prim function with the graph and a start node
mst = prim(graph, "A")

# Printing the minimum spanning tree
print("The minimum spanning tree is:")
for src, dest, weight in mst:
  print(f"{src} - {dest} : {weight}")
